test/Microsoft.Health.Fhir.Shared.Tests.E2E/Rest/HistoryTests.cs

```diff
@@ -194,52 +194,78 @@ public async Task GivenAValueForSinceAndBeforeWithModifications_WhenGettingSyste
+                    newResources.Add(await CreateResourceWithTag(Samples.GetJsonSample("Patient-f001").ToPoco(), tag));
+                    newResources.Add(await CreateResourceWithTag(Samples.GetJsonSample("Condition-For-Patient-f001").ToPoco(), tag));
 
-            var sinceUriString = HttpUtility.UrlEncode(since.ToString("o"));
+                    var sinceUriString = HttpUtility.UrlEncode(since.ToString("o"));
 
-            // Query all the recent changes
-            using FhirResponse<Bundle> allChanges = await _client.SearchAsync($"_history?_tag={tag}&_since=" + sinceUriString);
+                    // Query all the recent changes
+                    using FhirResponse<Bundle> allChanges = await _client.SearchAsync($"_history?_tag={tag}&_since=" + sinceUriString);
 
-            AssertCount(7, allChanges.Resource.Entry);
+                    AssertCount(7, allChanges.Resource.Entry);
 
-            // now choose a value of before that is as close as possible to one of the last updated times
-            var lastUpdatedTimes = allChanges.Resource.Entry.Select(e => e.Resource.Meta.LastUpdated).OrderBy(d => d.Value);
+                    // now choose a value of before that is as close as possible to one of the last updated times
+                    var lastUpdatedTimes = allChanges.Resource.Entry.Select(e => e.Resource.Meta.LastUpdated).OrderBy(d => d.Value);
 
-            var before = lastUpdatedTimes.ToList()[4];
-            var beforeUriString = HttpUtility.UrlEncode(before.Value.ToString("o"));
-            Thread.Sleep(500);
-            var firstSet = await _client.SearchAsync($"_history?_tag={tag}&_since={sinceUriString}&_before={beforeUriString}");
+                    var before = lastUpdatedTimes.ToList()[4];
+                    var beforeUriString = HttpUtility.UrlEncode(before.Value.ToString("o"));
+                    await Task.Delay(500);
+                    var firstSet = await _client.SearchAsync($"_history?_tag={tag}&_since={sinceUriString}&_before={beforeUriString}");
 
-            AssertCount(4, firstSet.Resource.Entry);
+                    AssertCount(4, firstSet.Resource.Entry);
 
-            sinceUriString = beforeUriString;
-            before = DateTime.UtcNow;
-            beforeUriString = HttpUtility.UrlEncode(before.Value.ToString("o"));
-            Thread.Sleep(500); // wait 500 milliseconds to make sure that the value passed to the server for _before is not a time in the future
-            var secondSet = await _client.SearchAsync($"_history?_tag={tag}&_since={sinceUriString}&_before={beforeUriString}");
+                    sinceUriString = beforeUriString;
+                    before = DateTime.UtcNow;
+                    beforeUriString = HttpUtility.UrlEncode(before.Value.ToString("o"));
+                    Thread.Sleep(500); // wait 500 milliseconds to make sure that the value passed to the server for _before is not a time in the future
```

Comment:
> Thread.Sleep(500); // wait 500 milliseconds to make sure that the value passed to the server for _before is not a time in the future [](http://example.com/codeflow?start=19&length=133)

I think this is a place where GetStartTimeForHistoryTest should be used, you can have more than 500ms time difference between local time and server time.

Reply:
GetStartTime makes some edits, and then gets a time after the last edit.  Here we want a timestamp which is very close to the current clock time, but after the edits made in the test case.  It is possible that we will have more than 500ms in difference, but since adding the 500ms delay, we have not seen the error that "before" is invalid due to being in the future.

Reply:
I'm still don't understand.

Here is my understanding.
We post resources to a server, one of the machine set lastupdated value to resource itself and it get converted to 
either ResourceSurrugateId or _lastUpdated field.
And it get stored.
As soon as it posted in persisted layer, it doesn't change.

You pick since and before timestamps. Lock them on testing machine, and send to a server, which convert them into parameters for search and filtering.

Neither values in persistent layer would be changed in this 500 ms, neither values you plan to send.

During history search, timeclock on machine which does search shouldn't matter.
There is no additional changes to resources after you create them, so there gonna be no new records.

What this delay try to accomplish? I truly don't understand.

Reply:
When we send the "before" string value to the server the server will compare the value that is received with the current clock on the server.  If the value received is in the future (from the servers point of view) then it will throw an error.  This 500ms delay tries to avoid that.  There is not guarantee that it will completely avoid the problem, but it seems to be good enough.  That is the purpose of the delay.

Reply:
Ok, I see, there is validation on server side based on parameters.
Can we not have it?
Why we should disallow user to search for data in future, considering how finnicky time works.

And on a side note, I don't see `_before` parameter been part of history search not in STU3([http://hl7.org/fhir/STU3/http.html#history](http://hl7.org/fhir/STU3/http.html#history))
not in R4 [https://www.hl7.org/fhir/http.html#history](https://www.hl7.org/fhir/http.html#history)
not in R5 [http://hl7.org/fhir/2020Sep/http.html#history](http://hl7.org/fhir/2020Sep/http.html#history)

Another side note, fact what we use web app time rather than database time is quite scary, I hope place where we host them is take care of time drift, otherwise you can have web app putting records in yesterday or even further in the past if it's time clock is in not in sync.

Reply:
Both good points.  I think you are correct, we could remove the check for the _before value, I don't think it would hurt anything, and we also should check on where we are getting our timestamps.  I added a user story to track them: 87440.